File: ios/DeviceOrientationModule.m
#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(DeviceOrientationModule, NSObject)

RCT_EXTERN_METHOD(lockToLandscape)
RCT_EXTERN_METHOD(lockToPortrait)
RCT_EXTERN_METHOD(unlockAllOrientations)
RCT_EXTERN_METHOD(isLocked:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

@end

File: android/src/main/java/com/kartikbhalla/orientation/DeviceOrientationModule.kt
package com.kartikbhalla.orientation

import android.content.pm.ActivityInfo
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.Promise
import com.facebook.react.module.annotations.ReactModule

@ReactModule(name = DeviceOrientationModule.NAME)
class DeviceOrientationModule(reactContext: ReactApplicationContext) :
  NativeOrientationSpec(reactContext) {

  private var currentLock: Int? = null

  companion object {
    const val NAME = "DeviceOrientation"
  }

  override fun getName(): String {
    return NAME
  }

  override fun lockToLandscape() {
    currentActivity?.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
    currentLock = ActivityInfo.SCREEN_ORIENTATION_LANDSCAPE
  }

  override fun lockToPortrait() {
    currentActivity?.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
    currentLock = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT
  }

  override fun unlockAllOrientations() {
    currentActivity?.requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED
    currentLock = null
  }

  override fun isLocked(promise: Promise) {
    promise.resolve(currentLock != null)
  }
}
